File: springboot/src/main/java/com/edufelip/meer/web/AuthController.java
package com.edufelip.meer.web;

import com.edufelip.meer.domain.repo.AuthUserRepository;
import com.edufelip.meer.domain.auth.*;
import com.edufelip.meer.dto.AuthDtos;
import com.edufelip.meer.dto.ProfileDto;
import com.edufelip.meer.mapper.AuthMappers;
import com.edufelip.meer.mapper.Mappers;
import com.edufelip.meer.security.token.InvalidRefreshTokenException;
import com.edufelip.meer.security.token.InvalidTokenException;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.Map;

@RestController
@RequestMapping("/auth")
public class AuthController {

    private final LoginUseCase loginUseCase;
    private final SignupUseCase signupUseCase;
    private final GoogleLoginUseCase googleLoginUseCase;
    private final AppleLoginUseCase appleLoginUseCase;
    private final RefreshTokenUseCase refreshTokenUseCase;
    private final ForgotPasswordUseCase forgotPasswordUseCase;
    private final AuthUserRepository authUserRepository;
    private final com.edufelip.meer.security.token.TokenProvider tokenProvider;

    public AuthController(LoginUseCase loginUseCase,
                          SignupUseCase signupUseCase,
                          GoogleLoginUseCase googleLoginUseCase,
                          AppleLoginUseCase appleLoginUseCase,
                          RefreshTokenUseCase refreshTokenUseCase,
                          ForgotPasswordUseCase forgotPasswordUseCase,
                          AuthUserRepository authUserRepository,
                          com.edufelip.meer.security.token.TokenProvider tokenProvider) {
        this.loginUseCase = loginUseCase;
        this.signupUseCase = signupUseCase;
        this.googleLoginUseCase = googleLoginUseCase;
        this.appleLoginUseCase = appleLoginUseCase;
        this.refreshTokenUseCase = refreshTokenUseCase;
        this.forgotPasswordUseCase = forgotPasswordUseCase;
        this.authUserRepository = authUserRepository;
        this.tokenProvider = tokenProvider;
    }

    @PostMapping("/login")
    public ResponseEntity<AuthDtos.AuthResponse> login(@RequestBody AuthDtos.LoginRequest body) {
        var result = loginUseCase.execute(body.email(), body.password());
        return ResponseEntity.ok(new AuthDtos.AuthResponse(result.getToken(), result.getRefreshToken(), AuthMappers.toDto(result.getUser())));
    }

    @PostMapping("/refresh")
    public ResponseEntity<AuthDtos.AuthResponse> refresh(@RequestBody AuthDtos.RefreshRequest body) {
        var result = refreshTokenUseCase.execute(body.refreshToken());
        return ResponseEntity.ok(new AuthDtos.AuthResponse(result.getToken(), result.getRefreshToken(), AuthMappers.toDto(result.getUser())));
    }

    @PostMapping("/forgot-password")
    public ResponseEntity<Map<String, String>> forgotPassword(@RequestBody AuthDtos.ForgotPasswordRequest body) {
        forgotPasswordUseCase.execute(body.email());
        return ResponseEntity.ok(Map.of("message", "Reset email sent"));
    }

    @PostMapping("/google")
    public ResponseEntity<AuthDtos.AuthResponse> loginWithGoogle(@RequestBody AuthDtos.GoogleLoginRequest body,
                                                                 @RequestHeader(name = "Authorization", required = false) String authHeader) {
        String idToken = body.idToken();
        if ((idToken == null || idToken.isBlank()) && authHeader != null && authHeader.startsWith("Bearer ")) {
            idToken = authHeader.substring("Bearer ".length()).trim();
        }
        if (idToken == null || idToken.isBlank()) throw new InvalidGoogleTokenException();
        var result = googleLoginUseCase.execute(idToken, body.client());
        return ResponseEntity.ok(new AuthDtos.AuthResponse(result.getToken(), result.getRefreshToken(), AuthMappers.toDto(result.getUser())));
    }

    @PostMapping("/apple")
    public ResponseEntity<AuthDtos.AuthResponse> loginWithApple(@RequestBody AuthDtos.AppleLoginRequest body) {
        if (!"apple".equalsIgnoreCase(body.provider())) throw new InvalidAppleTokenException();
        var result = appleLoginUseCase.execute(body.idToken(), body.authorizationCode(), body.client());
        return ResponseEntity.ok(new AuthDtos.AuthResponse(result.getToken(), result.getRefreshToken(), AuthMappers.toDto(result.getUser())));
    }

    @PostMapping("/signup")
    public ResponseEntity<AuthDtos.AuthResponse> signup(@RequestBody AuthDtos.SignupRequest body) {
        var result = signupUseCase.execute(body.name(), body.email(), body.password());
        return ResponseEntity.status(HttpStatus.CREATED)
                .body(new AuthDtos.AuthResponse(result.getToken(), result.getRefreshToken(), AuthMappers.toDto(result.getUser())));
    }

    @GetMapping("/me")
    public ResponseEntity<Map<String, Object>> me(@RequestHeader("Authorization") String authHeader) {
        String token = extractBearer(authHeader);
        var payload = tokenProvider.parseAccessToken(token);
        var user = authUserRepository.findById(payload.getUserId()).orElseThrow(InvalidTokenException::new);
        ProfileDto profile = Mappers.toProfileDto(user, true);
        return ResponseEntity.ok(Map.of("user", profile));
    }

    @ExceptionHandler(InvalidCredentialsException.class)
    public ResponseEntity<Map<String, String>> handleInvalidCredentials(InvalidCredentialsException ex) {
        return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body(Map.of("message", ex.getMessage()));
    }

    @ExceptionHandler(EmailAlreadyRegisteredException.class)
    public ResponseEntity<Map<String, String>> handleEmailTaken(EmailAlreadyRegisteredException ex) {
        return ResponseEntity.status(HttpStatus.CONFLICT).body(Map.of("message", ex.getMessage()));
    }

    @ExceptionHandler({InvalidGoogleTokenException.class, InvalidAppleTokenException.class, InvalidTokenException.class, InvalidRefreshTokenException.class})
    public ResponseEntity<Map<String, String>> handleInvalidToken(RuntimeException ex) {
        return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body(Map.of("message", ex.getMessage()));
    }

    private String extractBearer(String authHeader) {
        if (authHeader == null || !authHeader.startsWith("Bearer ")) throw new InvalidTokenException();
        return authHeader.substring("Bearer ".length()).trim();
    }
}

File: springboot/src/main/kotlin/com/edufelip/meer/web/AuthController.kt
package com.edufelip.meer.web

import com.edufelip.meer.domain.auth.InvalidCredentialsException
import com.edufelip.meer.domain.auth.LoginUseCase
import com.edufelip.meer.domain.auth.SignupUseCase
import com.edufelip.meer.domain.auth.EmailAlreadyRegisteredException
import com.edufelip.meer.domain.auth.GoogleLoginUseCase
import com.edufelip.meer.domain.auth.InvalidGoogleTokenException
import com.edufelip.meer.domain.auth.RefreshTokenUseCase
import com.edufelip.meer.domain.auth.AppleLoginUseCase
import com.edufelip.meer.domain.auth.InvalidAppleTokenException
import com.edufelip.meer.domain.auth.ForgotPasswordUseCase
import com.edufelip.meer.domain.AuthUserRepository
import com.edufelip.meer.dto.AuthResponse
import com.edufelip.meer.dto.LoginRequest
import com.edufelip.meer.dto.SignupRequest
import com.edufelip.meer.dto.GoogleLoginRequest
import com.edufelip.meer.dto.RefreshRequest
import com.edufelip.meer.dto.AppleLoginRequest
import com.edufelip.meer.dto.ForgotPasswordRequest
import com.edufelip.meer.mapper.toDto
import com.edufelip.meer.security.token.InvalidTokenException
import com.edufelip.meer.security.token.TokenProvider
import com.edufelip.meer.security.token.InvalidRefreshTokenException
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.RequestHeader
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/auth")
class AuthController(
    private val loginUseCase: LoginUseCase,
    private val signupUseCase: SignupUseCase,
    private val googleLoginUseCase: GoogleLoginUseCase,
    private val appleLoginUseCase: AppleLoginUseCase,
    private val refreshTokenUseCase: RefreshTokenUseCase,
    private val forgotPasswordUseCase: ForgotPasswordUseCase,
    private val tokenProvider: TokenProvider,
    private val authUserRepository: AuthUserRepository
) {

    @PostMapping("/login")
    fun login(@RequestBody body: LoginRequest): ResponseEntity<AuthResponse> {
        val result = loginUseCase.execute(body.email, body.password)
        return ResponseEntity.ok(
            AuthResponse(
                token = result.token,
                refreshToken = result.refreshToken,
                user = result.user.toDto()
            )
        )
    }

    @PostMapping("/refresh")
    fun refresh(@RequestBody body: RefreshRequest): ResponseEntity<AuthResponse> {
        val result = refreshTokenUseCase.execute(body.refreshToken)
        return ResponseEntity.ok(
            AuthResponse(
                token = result.token,
                refreshToken = result.refreshToken,
                user = result.user.toDto()
            )
        )
    }

    @PostMapping("/forgot-password")
    fun forgotPassword(@RequestBody body: ForgotPasswordRequest): ResponseEntity<Map<String, String>> {
        forgotPasswordUseCase.execute(body.email)
        return ResponseEntity.ok(mapOf("message" to "Reset email sent"))
    }

    @PostMapping("/google")
    fun loginWithGoogle(
        @RequestBody body: GoogleLoginRequest,
        @RequestHeader(name = "Authorization", required = false) authHeader: String?
    ): ResponseEntity<AuthResponse> {
        val idToken = body.idToken ?: extractBearerOptional(authHeader) ?: throw InvalidGoogleTokenException()
        val result = googleLoginUseCase.execute(idToken, body.client)
        return ResponseEntity.ok(
            AuthResponse(
                token = result.token,
                refreshToken = result.refreshToken,
                user = result.user.toDto()
            )
        )
    }

    @PostMapping("/apple")
    fun loginWithApple(@RequestBody body: AppleLoginRequest): ResponseEntity<AuthResponse> {
        if (body.provider.lowercase() != "apple") throw InvalidAppleTokenException()
        val result = appleLoginUseCase.execute(body.idToken, body.authorizationCode, body.client)
        return ResponseEntity.ok(
            AuthResponse(
                token = result.token,
                refreshToken = result.refreshToken,
                user = result.user.toDto()
            )
        )
    }

    @PostMapping("/signup")
    fun signup(@RequestBody body: SignupRequest): ResponseEntity<AuthResponse> {
        val result = signupUseCase.execute(body.name, body.email, body.password)
        return ResponseEntity.status(HttpStatus.CREATED).body(
            AuthResponse(
                token = result.token,
                refreshToken = result.refreshToken,
                user = result.user.toDto()
            )
        )
    }

    @GetMapping("/me")
    fun me(@RequestHeader("Authorization") authHeader: String?): ResponseEntity<Map<String, Any?>> {
        val token = extractBearer(authHeader)
        val payload = tokenProvider.parseAccessToken(token)
        val user = authUserRepository.findById(payload.userId).orElseThrow { InvalidTokenException() }
        val userDto = user.let { com.edufelip.meer.dto.UserDto(id = it.id.toString(), name = it.displayName, email = it.email) }
        return ResponseEntity.ok(mapOf("user" to userDto))
    }

    private fun extractBearer(header: String?): String {
        if (header == null || !header.startsWith("Bearer ")) throw InvalidTokenException()
        return header.removePrefix("Bearer ").trim()
    }

    private fun extractBearerOptional(header: String?): String? {
        if (header == null || !header.startsWith("Bearer ")) return null
        return header.removePrefix("Bearer ").trim()
    }

    @ExceptionHandler(InvalidCredentialsException::class)
    fun handleInvalidCredentials(ex: InvalidCredentialsException): ResponseEntity<Map<String, String>> {
        return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body(mapOf("message" to ex.message!!))
    }

    @ExceptionHandler(EmailAlreadyRegisteredException::class)
    fun handleEmailTaken(ex: EmailAlreadyRegisteredException): ResponseEntity<Map<String, String>> {
        return ResponseEntity.status(HttpStatus.CONFLICT).body(mapOf("message" to ex.message!!))
    }

    @ExceptionHandler(InvalidTokenException::class)
    fun handleInvalidToken(ex: InvalidTokenException): ResponseEntity<Map<String, String>> {
        return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body(mapOf("message" to ex.message!!))
    }

    @ExceptionHandler(InvalidGoogleTokenException::class)
    fun handleInvalidGoogleToken(ex: InvalidGoogleTokenException): ResponseEntity<Map<String, String>> {
        return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body(mapOf("message" to ex.message!!))
    }

    @ExceptionHandler(InvalidRefreshTokenException::class)
    fun handleInvalidRefresh(ex: InvalidRefreshTokenException): ResponseEntity<Map<String, String>> {
        return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body(mapOf("message" to ex.message!!))
    }

    @ExceptionHandler(InvalidAppleTokenException::class)
    fun handleInvalidApple(ex: InvalidAppleTokenException): ResponseEntity<Map<String, String>> {
        return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body(mapOf("message" to ex.message!!))
    }
}
